### PR TITLE
Remove incorrect `fmt_expression ~epi` mechanism

### PR DESCRIPTION
**Pull request by Julow**
Fix https://github.com/ocaml-ppx/ocamlformat/issues/2406

This argument doesn't work similarly to the `~epi` argument of other functions. It meant different things depending on the expression:

- `Pexp_match` and `Pexp_apply`: it was similar to `~pro`.
- `Pexp_constant`: like the usual `~epi` but was passed a break.
- all other cases: it was not used.

These rules are now handled in `fmt_args_grouped`, the only place where they made sense. The `~epi` argument is removed.

In the first case, the `~pro` argument is used instead. This requires refactoring `fmt_expression` to make sure that every cases use `~pro` exactly once.

### CODE REVIEW

File: lib/Fmt_ast.ml [hunk 1]
```diff
@@ -1714,19 +1724,20 @@ and fmt_expression c ?(box = true) ?pro ?epi ?eol ?parens ?(indent_wrap = 0)
...
-                  ( str "%"
-                  $ hovbox 2
-                      ( fmt_str_loc c name $ str " fun "
-                      $ fmt_attributes c ~suf:" " call.pexp_attributes
-                      $ fmt_fun_args c xargs $ fmt_opt fmt_cstr $ fmt "@ ->"
-                      )
-                  $ fmt "@ " $ fmt_expression c xbody ) )
-           $ fmt "@ ;@ "
-           $ list grps " ;@;<1000 0>" fmt_grp ) )
+      pro
```
Comment (Julow) [inline lib/Fmt_ast.ml:1713]: Ideally the `pro` would be pushed as far as possible into the boxes (and perhaps the opening parenthese should be appended to the `pro`) and the outer box should be removed.
But this is out of the scope of this PR.

File: lib/Fmt_ast.ml [hunk 2]
```diff
@@ -1454,15 +1458,22 @@ and fmt_args_grouped ?epi:(global_epi = noop) c ctx args =
...
+      | _ -> None
+    in
     let epi =
-      match (lbl, last) with
-      | _, true -> None
-      | Nolabel, _ -> Some (fits_breaks "" ~hint:(1000, -1) "")
-      | _ -> Some (fits_breaks "" ~hint:(1000, -3) "")
+      match ast.pexp_desc with
+      | Pexp_constant _ when not last -> fits_breaks "" ~hint:(1000, -2) ""
+      | _ -> noop
```
Comment (Julow) [inline lib/Fmt_ast.ml:1458]: I found out that the first rule was useless. Only this one remains.

The `break` code below could probably handle un-grouping `Pexp_constant` better. This introduces hard to control indentation.

Comment (Julow) [reply lib/Fmt_ast.ml]: This rule is actually more complex than that: it doesn't apply to every kinds of constants, when `break-string-literals=never` and in other cases that I couldn't find.

The regression are down to +126 -80 and are easily explained: Always break after multi-line string argument.

**Comment by Julow** [comment]
Test branch shows a new regression:
```diff
diff --git a/stdlib/marshal.mli b/stdlib/marshal.mli
index 507af732b3..7aa6859768 100644
--- a/stdlib/marshal.mli
+++ b/stdlib/marshal.mli
@@ -53,7 +53,8 @@ type extern_flags =
   | No_sharing  (** Don't preserve sharing *)
   | Closures  (** Send function closures *)
   | Compat_32  (** Ensure 32-bit compatibility *)
-  | Compression  (** Compress the output if possible
+  | Compression
+      (** Compress the output if possible
                      @since 5.1 *)
 
 val to_channel : out_channel -> 'a -> extern_flags list -> unit
```

File: lib/Fmt_ast.ml (hunk 2 above)
Comment (Julow) [reply lib/Fmt_ast.ml]: The regressions are due to hard to control implicit rules. I suggest making these explicit in https://github.com/ocaml-ppx/ocamlformat/pull/2448 (by introducing more regressions but on purpose)

Comment (Julow) [reply lib/Fmt_ast.ml]: This change is merged in https://github.com/ocaml-ppx/ocamlformat/pull/2453 and this PR has no more regressions!

**Comment by hhugo** [comment]
You should add a regression test for #2416

**Comment by Julow** [comment]
Good idea, added.

**Comment by Julow** [comment]
I think this is good to merge.

**Pull request merged**